torch/distributed/tensor/_random.py

```diff
@@ -260,23 +255,48 @@ def _set_pre_op_offset(self, spec: DTensorSpec) -> None:
         """
         dtensor_shape = spec.shape
         mesh = spec.mesh
-        dim_map = spec.dim_map
+        # note: dim_map does not allow double sharding which is the FSDP(fully_shard)+TP
+        # case. Replace the custom logic with dim_map once we support it.
+        dim_map: List[Union[int, List[int]]] = [-1] * spec.ndim
+        for i, placement in enumerate(spec.placements):
+            if isinstance(placement, Shard):
+                shard_dim = placement.dim
+                if dim_map[shard_dim] == -1:
+                    dim_map[shard_dim] = [i]
+                else:
+                    mesh_dim_list = dim_map[shard_dim]
+                    assert isinstance(mesh_dim_list, List)
+                    mesh_dim_list.append(i)
 
         # Compute shard coordinate:
         # The coordinate on each tensor dim is a tuple (idx, range)
         # If a DTensor is partitioned on its dim i into n shards, and the current rank
         # holds the j-th, then its shard coordinate will be (idx=j, range=n) on dim i
-        coordinate = mesh.get_coordinate()
-        assert coordinate is not None
-        shard_coord = [
-            coordinate[mesh_dim] if mesh_dim >= 0 else 0 for mesh_dim in dim_map
-        ]
-        shard_size = [
-            mesh.size(mesh_dim) if mesh_dim >= 0 else 1 for mesh_dim in dim_map
-        ]
+        mesh_coordinate = mesh.get_coordinate()
+        assert mesh_coordinate is not None
+        mesh_size = mesh.shape
+        shard_idx_by_dim = []
+        total_num_shards_by_dim = []  # total number of shards on each tensor dim
+        for mesh_dim in dim_map:
+            shard_idx = 0
+            total_num_shards = 1
+            # the tensor dim is sharded on more than 1 mesh dim
+            if isinstance(mesh_dim, List):
+                assert isinstance(mesh_dim, List)
+                rank_coord = [mesh_coordinate[d] for d in mesh_dim]
+                num_shards = [mesh_size[d] for d in mesh_dim]
+                # compute the shard idx and total number of shards
+                for idx, size in zip(rank_coord, num_shards):
+                    shard_idx = shard_idx * size + idx
+                    total_num_shards *= size
+
+            shard_idx_by_dim.append(shard_idx)
```

Comment:
ah, so we make a flattened 'shard_idx' for the cases where we have a tensor dim sharded over multiple mesh dims.  so instead of having 'shard idx = (1,0)' we get 'shard idx=3'